source/mesh_deformation/interface.cc

```diff
@@ -722,6 +740,253 @@ namespace aspect
 
 
 
+    template <int dim>
+    void MeshDeformationHandler<dim>::compute_mesh_displacements_gmg()
+    {
+      // Same as compute_mesh_displacements, but using GMG instead of AMG.
+
+      // We use this gmg solver only when the gmg stokes solver is used
+      // for the following reasons (TODO):
+      // 1. this gmg solver does not support periodic boundary conditions
+      // 2. To use this solver even when gmg stokes solver is not used, we need to
+      //    initialize the triangulation with Triangulation<dim>::limit_level_difference_at_vertices
+      //    and parallel::distributed::Triangulation<dim>::construct_multigrid_hierarchy
+      // 3. Although this gmg solver is much faster than the amg solver, it's only tested for
+      //    limited free surface cases.
+
+      TimerOutput::Scope timer (this->sim.computing_timer, "gmg displacement");
+
+      Assert(mesh_deformation_fe.degree == 1, ExcNotImplemented());
+      const unsigned int mesh_deformation_fe_degree = 1;
+
+      using SystemMatrixType = dealii::MatrixFreeOperators::
+                               LaplaceOperator<dim, mesh_deformation_fe_degree, mesh_deformation_fe_degree + 1, dim>;
+
+      SystemMatrixType mf_mesh_matrix;
+
+      MGLevelObject<SystemMatrixType> mg_matrices;
+
+      typename MatrixFree<dim, double>::AdditionalData additional_data;
+      additional_data.tasks_parallel_scheme =
+        MatrixFree<dim, double>::AdditionalData::none;
+      UpdateFlags update_flags = UpdateFlags(update_values | update_JxW_values | update_gradients);
+      additional_data.mapping_update_flags = update_flags;
+      std::shared_ptr<MatrixFree<dim, double>> system_mf_storage(
+                                              new MatrixFree<dim, double>());
+      system_mf_storage->reinit(*sim.mapping,
+                                mesh_deformation_dof_handler,
+                                mesh_velocity_constraints,
+                                QGauss<1>(mesh_deformation_fe_degree + 1),
+                                additional_data);
+      mf_mesh_matrix.initialize(system_mf_storage);
+
+      // correct rhs:
+      dealii::LinearAlgebra::distributed::Vector<double> u0, corrected_rhs, mf_solution;
+      mf_mesh_matrix.initialize_dof_vector(u0);
+      mf_mesh_matrix.initialize_dof_vector(corrected_rhs);
+      mf_mesh_matrix.initialize_dof_vector(mf_solution);
+      u0 = 0.;
+      mesh_velocity_constraints.distribute(u0);
+      u0.update_ghost_values();
+
+      corrected_rhs = 0.;
+
+      FEEvaluation<dim, mesh_deformation_fe_degree, mesh_deformation_fe_degree + 1, dim, double> phi(*mf_mesh_matrix.get_matrix_free());
```

Comment:
I know that this is syntax in line with examples in deal.II, but in the ASPECT matrix free code the FEEvaluation objects are named after their solution variable (compare stokes_matrix_free.cc:733). I would suggest to follow this here too, i.e. `phi` -> `mesh_deformation`.